docs/design/sql/20-sql-partition-pruning.md

```diff
@@ -0,0 +1,338 @@
+where `partColumnMExpr` is in form
+`partColumnM <operator> <arguments>`
+where `operator` is one of the following operators:
+- `=`
+- `SEARCH` with `Sarg`
+- `BETWEEN` for integer types (TODO: if we want to support this in the first version, makes sense only for reasonably small range)
+- TODO: `IS NULL`?
+
+and operator `arguments` are:
+- literal
+- query parameter
+- complex expression using only the above `arguments` and any deterministic functions
+  (in particular references to columns are not allowed)
+
+and `residualFilter` is any expression, in particular it can be `TRUE`.  `residualFilter` may reference any columns,
+also the partitioning columns.
+
+Clarifications:
+- range partitioning (eg. `order_date between day1 and day1`) currently is not supported,
+  only hash/equality based partitioning is supported.
+- each `OR` branch in `E` might use different partitioning key. Currently, such case is not supported.
+- `partCol1 = partCol2 AND partCol2 = constantX` is not partition-complete, but can be transformed to such form
+  by inlining constant: `partCol1 = constantX AND partCol2 = constantX`. Such transformations are out-of-scope
+  for this TDD, but may be performed now or in future by the SQL optimizer independently.
+
+#### Goals
+
+- Improve performance of queries using IMap with `AttributePartitioningStrategy` which use attributes defined in it
+- Improve performance of queries using IMap `__key` which are not eligible to be converted to `IMapSelectPlan`
+  and have to create a Jet job
+
+#### Non-goals
+
+- Support partition pruning for other sources than IMap (can be considered in the future)
+- Complex expression transformations of filters to extract partitioning information
+  (for complex cases users will have to write their predicates in supported form)
+- Tracking of functional dependencies between attributes
+- Building a `__key` from constituent attributes, eg. if `__key` has 2 attributes, we could convert
+  `__key.a1=X and __key.a2=Y` to `__key=(X,Y)`. 
+  TODO: maybe we already do that?
+
+### Technical Design
+
+#### Assumptions
+
+##### Equi-partitions
+
+For IMap we support equality-based partitions which is inline with IMap being a hash table.
+Range-partitioning (eg. for date ranges) will be not supported.
+
+#### Partitioning metadata available to SQL optimizer
+
+Calcite SQL optimizer need to know how IMap is partitioned.
+
+List of columns comprising `AttributePartitioningStrategy` of the IMap will be available in `PartitionedMapTable`.
+(note that `__key` is a special case that should be handled with either an extension to `AttributePartitioningStrategy`
+or in a separate specialized Partitioning Strategy that will explicitly ignore case when whole key object implements 
+`PartitionAware` interface).
+
+#### Information if RelNode is partitioned and how
+
+For some/all RelNodes we will be using information about how its input(s) are partitioned:
```

Comment:
I feel this chapter is lack of description, but it is mostly, the main chapter in this TDD.